xbindkeys.1
.TH XBindKeys 1 "Tue Apr 19 2014"
.UC 1
.SH "NAME"
xbindkeys \- a grabbing keys program for X
.SH "SYNOPSIS"
\fBxbindkeys\fP [ \fIoptions\fP ]
.br
.SH "DESCRIPTION"
\fIXbindKeys\fP is a program that grabs key and mouse button events
in X and starts associated shell commands.

.LP
.SH "COMMAND LINE OPTIONS"
Available command line options are as follows:
.TP
\fB-d, --defaults\fP
Print a default rc file
.TP
\fB-f, --file\fP
Use an alternative rc file
.TP
\fB-h, --help\fP
Short help on options plus version/author info.
.TP
\fB-X, --display\fP
Set X display to use
.TP
\fB-v, --verbose\fP
More information on xbindkeys when it runs

.TP
\fB-k, --key\fP
Identify one key pressed (useful for configuration)

.TP
\fB-mk, --multikey\fP
Identify multi key pressed (useful for configuration)

.TP
\fB-g, --geometry\fP
size and position of window with -k|-mk option

.TP
\fB-n, --nodaemon\fP
don't start as daemon

.TP
\fB-y N, --delay N\fP
Delay N milliseconds before issuing command.

.LP
If guile support has not been disabled:
.TP
\fB-dg, --defaults-guile\fP
Print a default guile configuration file

.TP
\fB-fg, --file-guile\fP
Use an alternative guile configuration file

.SH "XBINDKEYSRC"
XBindKeys default configuration file is at $HOME/.xbindkeysrc,
used only for one user.
.LP
The configuration file consists of commands in double quotes,
and associated keys in the next line.
.LP
Comments are started with a pound (#) and extend to the end of the line.
.LP
You can see a default configuration with --defaults
.LP
.fi

.SH "KEYSYMS"
The program expects combinations of keysyms to be used by using plus (+)
characters to combine them together.
.LP
The format of a command line is:
.nf
    "command to start &"
.nf
       associated key
.LP
or, if using a condition (described below)
.nf
    "command to start &"
.nf
    %condition type target
.nf
       associated key
.LP
.fi
To specify an associated key, you can use `xbindkeys --key` or
 `xbindkeys --multikey` and put one of the two lines in the
configuration file.
.LP
A list of keys is in /usr/include/X11/keysym.h and in
/usr/include/X11/keysymdef.h.
The XK_ is not needed.
.LP
List of modifiers:
.nf
    Release, Control, Shift, Mod1 (Alt), Mod2 (NumLock),
.nf
    Mod3 (CapsLock), Mod4, Mod5 (Scroll).
.LP
.fi
The release modifier is not a standard X modifier, but you can
use it if you want to catch release events instead of press events.
Modifiers are not case sensitive.
.LP
By defaults, xbindkeys does not pay attention to the modifiers
NumLock, CapsLock and ScrollLock.
Add the lines below in the configuration file if you want to pay
attention to them.
.LP
    keystate_numlock = enable
.nf
    keystate_capslock = enable
.nf
    keystate_scrolllock= enable

.SH "USE OF DELAY"
A key can be changed by invoking \fIxdotool\fP as the command.
However, for this to work a small delay is often required before running
the command, otherwise the \fIxbindkeys\fP program sees the key.
The delay can be added in the control file with a line such as
.br
    delay = \fINNN\fP
.br
where \fINNN\fP is the number of milliseconds to delay.  A value of
100 is often sufficient if the key is released quickly afterwards.

The -y (or --delay) option can be used to test various values if the
delay line is not used.

.SH "CONDITIONS"
.LP
A condition line has the format
.nf
    %qualifier type target1
.fi    
and if present appears between the command line and the associated key line.
.PP
The \fIqualifier\fP is either WHEN or EXCLUDE.

The \fItype\fP is used to derive a string from the window where the event originated from.
The derived string is:
.nf
* the window name (WM_NAME property) if \fItype\fP is \fBNAME\fP;
.nf
* the window class (WM_CLASS property) if \fItype\fP is \fBCLASS\fP;
.nf
* the window's command (WM_COMMAND property) if \fItype\fP is \fBCOMMAND\fP.
.fi

The \fItarget\fP is an extended regexp (in the style of egrep).

The condition is true if
.nf
* the qualifier is \fBWHEN\fP and there is a match, or
.nf
* the qualifier is \fBEXCLUDE\fP and there is no match.
.fi

If the condition is true then the comand is executed.
.SH "WHEN THE CONDITION IS FALSE"
The behaviour when the condition evalues to false is controlled
by the presence of either of these lines in the rc file:
.in +4n
fallback = discard
.nf
fallback = pass
.in
Specifying \fBdiscard\fP means the event is ignored.
.nf
Specifying \fBpass\fP means the event is passed to the window as normal
(as if xkeybindings was not in effect).

The default is \fBdiscard\fP.
.SS Caution When Using Pass
Specifying fallback = pass causes a synchronous grab on the keyboard and mouse
(as opposed to an asynchrous grab when using discard).  Attempts to debug such a program
may lead to the keyboard or pointer being locked up.
.SH "EXAMPLES"
.LP
# control+shift+q starts xterm
.nf
"xterm"
.nf
  control+shift + q
.LP
# Menu key starts xbindkeys_show
.nf
"xbindkeys_show"
.nf
  Menu
.LP
# Control + mouse button 1 starts xterm
.nf
"xterm"
.nf
  Control + b:1
.LP
# Control+Shift+a  release event starts rxvt
.nf
"rxvt"
.nf
  release+control+shift + a
.LP
# Control + mouse button 2 release event starts rxvt
.nf
"rxvt"
.nf
  Control + b:2 + Release
.PP
# Use xtodotool to turn control-b into a Left key in a firefox window.
.nf
"xdotool key Left"
.nf
%WHEN FILE firefox
.nf
    control + b

.fi
.SH "SCHEME CONFIGURATION STYLE"
A scheme configuration can be used if guile support has not been disabled.
The standard configuration is loaded only if the scheme configuration is not
available.
.LP
For more details, please see here
.nf
  http://www.gnu.org/software/guile/guile.html
.LP
\fBNote: The guile configuration file is prefered for non trivial
configurations, for example double click, timed click or key combinations.\fP
.LP
The default location is ~/.xbindkeysrc.scm

.LP
Use xbindkeys --defaults-guile for more details.
.LP
And xbindkeys --defaults-guile > ~/.xbindkeysrc.scm to use this method.

.SH "NOTE"
xbindkeys reloads the configuration file if modified.
Send a HUP signal to force reloading:
.LP
killall -HUP xbindkeys
.LP
When a HUP signal is sent to xbindkeys, changes take effect only after
moving the mouse.

.SH "BUGS"
If you find a bug, please send mail to <xbindkeys-devel@nongnu.org>
.LP
.SH "HOMEPAGE"
.nf
http://www.nongnu.org/xbindkeys/
.fi

.LP
.SH "AUTHOR"
Philippe Brochard <hocwp@free.fr>
.nf
Marcello Mathias Herreshoff for guile support <marcello@hummer.stanford.edu>

.LP
.SH "FILES"
.nf
$HOME/.xbindkeysrc
        The users configuration file.
.LP
$HOME/.xbindkeysrc.scm
        The users configuration file in scheme
        (if guile support has not been disabled).
.fi

.LP
.SH "SEE ALSO"
.nf

\fBxbindkeys_show\fP
    Utility for showing the actual keybinding with xbindkeys
\fBxmodmap(1x)\fP
    Utility for modifying keymap & button mappings in X.
\fBxev(1x)\fP
    Print contents of X events.
\fB/usr/include/X11/keysymdef.h\fP
    X11 KeySym definitions.



ChangeLog
2022-02-14  Richard Sharman  <rsharman@pobox.com>

	* Added conditions:   The rc and scm files may now specify a
	condition required to run the command.  If the condition is satisfied,
	the command is run; if not by default the action is ignored but a global
	option allows the event to be passed to the window.

	* Added option to sleep before running command;  this is often
	required when xdotool is run.

	* Fixed some compilation warnings

	* Updated configue to check for guile 3

2020-05-23  Philippe Brochard <hocwp@free.fr>

	* Fix some compilation warnings

2020-05-23 Jason Miller  <jasom@fastmail.com>

	* Prevent scheme functions to be garbage collected

2011-03-09  Hans-Peter Deifel <hpdeifel@gmx.de>

	* options.c: Replace some deprecated guile functions.
	The GH interface has been removed in guile-2.0 so do not use it
	anymore in xbindkeys.

2011-02-21  Philippe Brochard  <pbrochard@common-lisp.net>

	* options.c (get_rc_guile_file): Fix a crash related to guile's GC
	and freeing of static memory. Thanks to Thomas Lindroth.

2010-12-04  Philippe Brochard  <pbrochard@common-lisp.net>

	* options.c: Fix some compilations warnings.

	* xbindkeys.c (adjust_display): adjust envstr_size to prevent a
	segfault with libX11-1.4.0.

2010-01-24  Alexander Clouter <alex@digriz.org.uk>

	* xbindkeys.c (event_loop): Doesn't stat on configuration file
	change the first time if poll-rc is false.

2010-01-23  Philippe Brochard  <hocwp@free.fr>
	* options.c: A poll-rc flag has been added to poll the rc/guile
	configuration file for update. By default no poll is done on
	configuration files. Many thanks to Alexander Clouter.

	* xbindkeys.c (event_loop): Stat both default rc file and guile rc
	file for reloading them when they change.

xbindkeys changelog by Philippe Brochard <hocwp@free.fr>


version:

1.8.3:
	Add a listen loop in the main loop. This fix a bug where, in the
	previous version, a keypress was needed to reload the
	configuration file after a keybinding change. Now one can edit the
	configuration file and keybindings are automatically updated after
	each save. Many thanks to Edwin Stang!

1.8.2:
	Correction of typos in the man page. An emphasis is made on the
	documentation about the guile configuration file (the prefered way
	for non trivial configuration).
	Many thanks to Brad Sawatzky.

1.8.1:
	Better Guile configuration with autotools.
	Minor compilation warning have been fixed.

1.8.0:
	Enable a full access to the xbindkeys internal from the guile
	scheme configuration file. A grabbed key can start a shell command
	or run a scheme function. This enable to do more powerfull things
	inside xbindkeys whitout the need of an external shell script
	(like double click, timed double click or keys combinations).

---------------------------------------------------------------------------

1.7.4:
	xbindkeys.spec: Change Copyright to License.
	Little english correction in warnings.
	Configuration files are reloaded on the fly when they have
	changed so there is no more need to send a HUP signal to reload
	them (Many thanks to Shish).
	Use `scm_c_define_gsubr' instead of 'scm_make_gsubr' with the
	guile mode.


1.7.3:
	Bugfix under Solaris, XBindKeys built with -DFORK_FLAG=1 fails to
	wait for processes it spawns, leaving lots of defunct processes.
	Thanks to Ciaran Anscomb for the patch.

1.7.2:
	Changes to grab keys in all display screens and not only in the
	default root window (Xinerama suport).
	Bugfix for the Guile support in configure.ac.

1.7.1:
	Verbose mode is not started as daemon.
	Check rc file before starting as daemon (so warning about
	configuration files are visible).

1.7.0:
	Add guile/scheme configurtion file style (Many thanks to Marcello
	Mathias Herreshoff).

---------------------------------------------------------------------------

1.6.5:
        Add a --disable-tk option to prevent xbindkeys_show install.
	So no tcl/tk dependency is needed.
	Add a NOTE in xbindkeys man page to say that xbindkeys can
	reread its config file with a -HUP signal (killall -HUP xbindkeys).
	Add two verbose test (forgotten in other version).

1.6.4:
	Add a --disable-fork option  in configure.ac to start one command
	at a time (use system instead of fork+exec).
	Remove umask (022) in daemonize function which cause troubles.

1.6.3:
	Change to switch to autoconf2.5X and automake >= 1.6
	Remove chdir ("/tmp")  in daemonize function.

1.6.2:
	Remove AC_PROG_RANLIB and AC_PROG_MAKE_SET from configure.in
	to prevent rebuit of Makefile when doing a make.

1.6.1:
	Minor buggfix : use strtol instead atoi for use with
	hexadecimal number.
	Complete daemonzie function

1.6.0:
	Add: Release event for keys and buttons.
	Don't use threads any more (use fork+exec instead)
	Better daemonize function
	Simplification of configure.in (better test for X)
	Don't use sscanf any more

---------------------------------------------------------------------------

1.5.5:
	Change: better identification of modifiers with --key and
	--mulitkey options.

1.5.4:
        Change: add warning when a key is unknown in .xbindkeysrc.

1.5.3:
	Bugfix: if the keycode is unknow then xbindkeys don't asign a
	key with keycode=0.
	Change: The default RC file don't use the Menu key witch cause
	problems.

1.5.2:
	Change: if there's a conflict with keys then xbindkeys just
	make a warning and continue with other keys.

1.5.1:
	Add: a new interface for xbindkeys_show (use scrollbar).

1.5.0:
	Add: xbindkeys have now the ability to catch Mouse Button Events.
	so: "xterm"
	      control + b:2
	starts an xterm when mouse button 2 and control are pressed.

---------------------------------------------------------------------------

1.4.1:
	Solve some minor problems when compiling with gcc 3.0.4 under GNU/Linux.

1.4.0:
	Change: rename all code to make it uniform.
	Add: start as daemon by default, but can be changed with
	  --nodaemon|-n option.
	Add: xbindkeys starts commands in threads, so they are not blocking

xbindkeys.c
/***************************************************************************
        xbindkeys : a program to bind keys to commands under X11.
                           -------------------
    begin                : Sat Oct 13 14:11:34 CEST 2001
    copyright            : (C) 2001 by Philippe Brochard
    email                : hocwp@free.fr
 ***************************************************************************/

/***************************************************************************
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 ***************************************************************************/

#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <signal.h>
#include <unistd.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <regex.h>
#include "xbindkeys.h"
#include "keys.h"
#include "options.h"
#include "get_key.h"
#include "grab_key.h"
#include "winfo.h"

#ifdef GUILE_FLAG
#include <libguile.h>
#endif

#include <X11/XKBlib.h>


void end_it_all (Display * d);

static Display *start (char *);
static void event_loop (Display *);
static int *null_X_error (Display *, XErrorEvent *);
static void reload_rc_file (void);
static void catch_HUP_signal (int sig);
static void catch_CHLD_signal (int sig);
static void start_as_daemon (void);


Display *current_display;  // The current display

extern char rc_file[512];
#ifdef GUILE_FLAG
extern char rc_guile_file[512];
#endif
extern int poll_rc;

#define SLEEP_TIME 100

#ifndef GUILE_FLAG
int
main (int argc, char **argv)
#else
int argc_t; char** argv_t;
void
inner_main (int argc, char **argv)
#endif
{
  Display *d;

#ifdef GUILE_FLAG
  argc = argc_t;
  argv = argv_t;
#endif

  get_options (argc, argv);

  if (!rc_file_exist ())
    exit (-1);

  if (have_to_start_as_daemon && !have_to_show_binding && !have_to_get_binding)
    start_as_daemon ();

  if (!display_name)
    display_name = XDisplayName (NULL);

  show_options ();

  d = start (display_name);
  current_display = d;

  if (detectable_ar)
    {
      Bool supported_rtrn;
      XkbSetDetectableAutoRepeat(d, True, &supported_rtrn);

      if (!supported_rtrn)
	{
	  fprintf (stderr, "Could not set detectable autorepeat\n");
	}
    }

  get_offending_modifiers (d);

  if (have_to_get_binding)
    {
      get_key_binding (d, argv, argc);
      end_it_all (d);
      exit (0);
    }

#ifdef GUILE_FLAG
  if (get_rc_guile_file () != 0)
    {
#endif
      if (get_rc_file () != 0)
	{
	  end_it_all (d);
	  exit (-1);
	}
#ifdef GUILE_FLAG
    }
#endif


  if (have_to_show_binding)
    {
      show_key_binding (d);
      end_it_all (d);
      exit (0);
    }

  grab_keys (d);

  /* This: for restarting reading the RC file if get a HUP signal */
  signal (SIGHUP, catch_HUP_signal);
  /* and for reaping dead children */
  signal (SIGCHLD, catch_CHLD_signal);

  if (verbose)
    printf ("starting loop...\n");
  event_loop (d);

  if (verbose)
    printf ("ending...\n");
  end_it_all (d);

#ifndef GUILE_FLAG
  return (0);			/* not reached... */
#endif
}

#ifdef GUILE_FLAG
int
main (int argc, char** argv)
{
  //guile shouldn't steal our arguments! we already parse them!
  //so we put them in temporary variables.
  argv_t = argv;
  argc_t = argc;
  //printf("Starting in guile mode...\n"); //debugery!
  scm_boot_guile(0,(char**)NULL,(void *)inner_main,NULL);
  return 0; /* not reached ...*/
}
#endif



static Display *
start (char *display)
{
  Display *d;
  int screen;

  d = XOpenDisplay (display);
  if (!d)
    {
      fprintf (stderr,
	       "Could not open display, check shell DISPLAY variable, \
and export or setenv it!\n");
      exit (1);
    }


  XAllowEvents (d, AsyncBoth, CurrentTime);

  for (screen = 0; screen < ScreenCount (d); screen++)
    {
      XSelectInput (d, RootWindow (d, screen),
		    KeyPressMask | KeyReleaseMask | PointerMotionMask);
    }

  return (d);
}



void
end_it_all (Display * d)
{
  ungrab_all_keys (d);

  close_keys ();
  XCloseDisplay (d);
}



static void
adjust_display (XAnyEvent * xany)
{
  size_t envstr_size = strlen(DisplayString(xany->display)) + 8 + 1;
  char* envstr = malloc ( (envstr_size + 2) * sizeof (char) );
  XWindowAttributes attr;
  char* ptr;
  char buf[16];

  snprintf (envstr, envstr_size, "DISPLAY=%s", DisplayString(xany->display));

  XGetWindowAttributes (xany->display, xany->window,  &attr);

  if (verbose)
    printf ("got screen %d for window %x\n", XScreenNumberOfScreen(attr.screen), (unsigned int)xany->window );

  ptr = strchr (strchr (envstr, ':'), '.');
  if (ptr)
    *ptr = '\0';

  snprintf (buf, sizeof(buf), ".%i", XScreenNumberOfScreen(attr.screen));
  strncat (envstr, buf, 16);

  putenv (envstr);
}



/* If this key has a condition, is it met? */
static int
allow_key (Display * d, XEvent *e, int i, Window cwindow)
{
  int match_required;
  int match_found;
  int condition_satisfied;
  char *string;
  int n;

  if (keys[i].condition.required < CONDITION_MATCH)
    return 1;	/* no condition or guile func, so honour the event */

  match_required = keys[i].condition.required == CONDITION_MATCH;
  switch (keys[i].condition.match)
    {
    case MATCH_COMMAND:
      string = window_command (d, cwindow);
      break;
    case MATCH_CLASS:
      string = window_class (d, cwindow);
      break;
    case MATCH_NAME:
      string = window_name (d, cwindow);
      break;
    default:
      fprintf (stderr, "Invalid condition type %d\n", keys[i].condition.match);
      exit (2);
    }
  if (string == NULL)
    {
      fprintf (stderr, "Could not find string for window 0x%lx\n", cwindow);
      return 1;         /* Arbitrary, but treat as no condition */
    }
  else
    {
      if (verbose)
	printf ("i=%d String for window 0x%lx is %s\n", i, cwindow, string);
      n = regexec (&keys[i].condition.regexp, string, 0, NULL, 0);
      match_found = n == 0;
      condition_satisfied = match_found == match_required;
      if (verbose)
	printf ("match_required: %d  match_found: %d  condition_satisfied: %d\n",
		match_required, match_found, condition_satisfied);
      XFree(string);
      return condition_satisfied;
    }
}

static void
swallow_event(Display * d, XEvent * ev, int event_mode)
{
    XAllowEvents (d, event_mode, ev->xkey.time);
}

static void
passthru_event(Display * d, XEvent * ev, int event_mode)
{
    /* pass it through as if we never intercepted it */
    XAllowEvents (d, event_mode, ev->xkey.time);
    XFlush (d); /* don't forget! */
}



static int
execute_key (Display * d, Keys_t * key, XEvent * e, Window client_window)
{
  current_windowid = client_window;	/* accessible from scheme */
  print_key (d, key);
  adjust_display (&(e->xany));
  return start_command_key (key, client_window);
}

static void
event_loop (Display * d)
{
  XEvent e;
  int i;
  struct stat rc_file_info;
  time_t rc_file_changed = 0;
  int event_is_for_us;
  int condition_was_true;
  Window cwindow;	/* client window */
#ifdef GUILE_FLAG
  time_t rc_guile_file_changed = 0;
  struct stat rc_guile_file_info;
#endif


  XSetErrorHandler ((XErrorHandler) null_X_error);

  if (poll_rc)
    {
      stat(rc_file, &rc_file_info);
      rc_file_changed = rc_file_info.st_mtime;
#ifdef GUILE_FLAG
      stat (rc_guile_file, &rc_guile_file_info);
      rc_guile_file_changed = rc_guile_file_info.st_mtime;
#endif
    }

  while (True)
    {
      while(poll_rc && !XPending(d))
	{
	  // if the rc file has been modified, reload it
	  stat (rc_file, &rc_file_info);
#ifdef GUILE_FLAG
	  // if the rc guile file has been modified, reload it
	  stat (rc_guile_file, &rc_guile_file_info);
#endif

	  if (rc_file_info.st_mtime != rc_file_changed
#ifdef GUILE_FLAG
	      || rc_guile_file_info.st_mtime != rc_guile_file_changed
#endif
	      )
	    {
	      reload_rc_file ();
	      if (verbose)
		{
		  printf ("The configuration file has been modified, reload it\n");
		}
	      rc_file_changed = rc_file_info.st_mtime;
#ifdef GUILE_FLAG
	      rc_guile_file_changed = rc_guile_file_info.st_mtime;
#endif
	    }

	  usleep(SLEEP_TIME*1000);
	}

      XNextEvent (d, &e);

      switch (e.type)
	{
	case KeyPress:
	  cwindow = client_window (d, e.xkey.subwindow);
	  if (verbose)
	    {
	      printf ("Key press !\n");
	      printf ("e.xkey.keycode=%d\n", e.xkey.keycode);
	      printf ("e.xkey.state=%d\n", e.xkey.state);
	      printf ("e.xkey.subwindow is 0x%lx\n", e.xkey.subwindow);
	      printf ("client window = 0x%lx\n", cwindow);
	    }

	  e.xkey.state &= ~(numlock_mask | capslock_mask | scrolllock_mask);

	  condition_was_true = 0;
	  for (i = 0; i < nb_keys; i++)
	    {
	      event_is_for_us = 0;
	      if (keys[i].type == SYM && keys[i].event_type == PRESS)
		{
		  if (e.xkey.keycode == XKeysymToKeycode (d, keys[i].key.sym)
		      && e.xkey.state == keys[i].modifier)
		    event_is_for_us = 1;
		}
	      else
	      if (keys[i].type == CODE && keys[i].event_type == PRESS)
		{
		  if (e.xkey.keycode == keys[i].key.code
		      && e.xkey.state == keys[i].modifier)
		    event_is_for_us = 1;
		}
	      if (event_is_for_us)
		{
		  if (allow_key(d, &e, i, cwindow))
		    if (execute_key(d, &(keys[i]), &e, cwindow))
		      condition_was_true = 1;
		}
	    }
	  if (condition_was_true)
	    {
	      swallow_event(d, &e, AsyncKeyboard);
	    }
	  else
	    {
	      passthru_event(d, &e, ReplayKeyboard);
	    }
	  break;

	case KeyRelease:
	  cwindow = client_window (d, e.xkey.subwindow);
	  if (verbose)
	    {
	      printf ("Key release !\n");
	      printf ("e.xkey.keycode=%d\n", e.xkey.keycode);
	      printf ("e.xkey.state=%d\n", e.xkey.state);
	      printf ("client window = 0x%lx\n", cwindow);
	    }

	  e.xkey.state &= ~(numlock_mask | capslock_mask | scrolllock_mask);

	  condition_was_true = 0;
	  for (i = 0; i < nb_keys; i++)
	    {
	      event_is_for_us = 0;
	      if (keys[i].type == SYM && keys[i].event_type == RELEASE)
		{
		  if (e.xkey.keycode == XKeysymToKeycode (d, keys[i].key.sym)
		      && e.xkey.state == keys[i].modifier)
		    event_is_for_us = 1;
		}
	      else
	      if (keys[i].type == CODE && keys[i].event_type == RELEASE)
		{
		  if (e.xkey.keycode == keys[i].key.code
		      && e.xkey.state == keys[i].modifier)
		    event_is_for_us = 1;
		}
	      if (event_is_for_us)
		{
		  if (allow_key(d, &e, i, cwindow))
		    if (execute_key(d, &(keys[i]), &e, cwindow))
		      condition_was_true = 1;
		}
	    }
	  if (condition_was_true)
	    {
	      swallow_event(d, &e, AsyncKeyboard);
	    }
	  else
	    {
	      passthru_event(d, &e, ReplayKeyboard);
	    }
	  break;

	case ButtonPress:
	  cwindow = client_window (d, e.xbutton.subwindow);
	  if (verbose)
	    {
	      printf ("Button press !\n");
	      printf ("e.xbutton.button=%d\n", e.xbutton.button);
	      printf ("e.xbutton.state=%d\n", e.xbutton.state);
	      printf ("client window = 0x%lx\n", cwindow);
	    }

	  e.xbutton.state &= 0x1FFF & ~(numlock_mask | capslock_mask | scrolllock_mask
			       | Button1Mask | Button2Mask | Button3Mask
			       | Button4Mask | Button5Mask);

	  condition_was_true = 0;
	  for (i = 0; i < nb_keys; i++)
	    {
	      if (keys[i].type == BUTTON && keys[i].event_type == PRESS)
		{
		  if (e.xbutton.button == keys[i].key.button
		      && e.xbutton.state == keys[i].modifier)
		    {
                      //printf("Replay!!!\n");
                      //ungrab_all_keys(d);
                      //XPutBackEvent(d, &e);
                      //sleep(1);
                      //grab_keys(d);
		      if (allow_key(d, &e, i, cwindow))
			{
			  if (execute_key(d, &(keys[i]), &e, cwindow))
			    condition_was_true = 1;
			}
		    }
		}
	    }
	  if (condition_was_true)
	    swallow_event(d, &e, AsyncPointer);
	  else
	    passthru_event(d, &e, ReplayPointer);
	  break;

	case ButtonRelease:
	  cwindow = client_window (d, e.xbutton.subwindow);
	  if (verbose)
	    {
	      printf ("Button release !\n");
	      printf ("e.xbutton.button=%d\n", e.xbutton.button);
	      printf ("e.xbutton.state=%d\n", e.xbutton.state);
	      printf ("client window = 0x%lx\n", cwindow);
	    }

	  e.xbutton.state &= 0x1FFF & ~(numlock_mask | capslock_mask | scrolllock_mask
			       | Button1Mask | Button2Mask | Button3Mask
			       | Button4Mask | Button5Mask);

	  condition_was_true = 0;
	  for (i = 0; i < nb_keys; i++)
	    {
	      if (keys[i].type == BUTTON && keys[i].event_type == RELEASE)
		{
		  if (e.xbutton.button == keys[i].key.button
		      && e.xbutton.state == keys[i].modifier)
		    {
		      if (allow_key(d, &e, i, cwindow))
			{
			  if (execute_key(d, &(keys[i]), &e, cwindow))
			    condition_was_true = 1;
			}

		    }
		}
	    }
	  if (condition_was_true)
	    swallow_event(d, &e, AsyncPointer);
	  else
	    passthru_event(d, &e, ReplayPointer);
	  break;

	default:
	  break;
	}
    }				/*  infinite loop */
}



static int *
null_X_error (Display * d, XErrorEvent * e)
{
  static int already = 0;

  /* The warning is displayed only once */
  if (already != 0)
    return (NULL);
  already = 1;

  printf ("\n*** Warning ***\n");
  printf ("Please verify that there is not another program running\n");
  printf ("which captures one of the keys captured by xbindkeys.\n");
  printf ("It seems that there is a conflict, and xbindkeys can't\n");
  printf ("grab all the keys defined in its configuration file.\n");

/*   end_it_all (d); */

/*   exit (-1); */

  return (NULL);
}



static void
reload_rc_file (void)
{
  int min, max;
  int screen;

  XDisplayKeycodes (current_display, &min, &max);

  if (verbose)
    printf ("Reload RC file\n");

  for (screen = 0; screen < ScreenCount (current_display); screen++)
    {
      XUngrabKey (current_display, AnyKey, AnyModifier, RootWindow (current_display, screen));
    }
  close_keys ();

#ifdef GUILE_FLAG
  if (get_rc_guile_file () != 0)
    {
#endif
      if (get_rc_file () != 0)
	{
	  end_it_all (current_display);
	  exit (-1);
	}
#ifdef GUILE_FLAG
    }
#endif

  grab_keys (current_display);
}


static void
catch_HUP_signal (int sig)
{
  reload_rc_file ();
}


static void
catch_CHLD_signal (int sig)
{
  pid_t child;

  /*   If more than one child exits at approximately the same time, the signals */
  /*   may get merged. Handle this case correctly. */
  while ((child = waitpid(-1, NULL, WNOHANG)) > 0)
    {
      if (verbose)
	printf ("Catch CHLD signal -> pid %i terminated\n", child);
    }
}





static void
start_as_daemon (void)
{
  pid_t pid;
  int i;

  pid = fork ();
  if (pid < 0)
    {
      perror ("Could not fork");
    }
  if (pid > 0)
    {
      exit (EXIT_SUCCESS);
    }

  setsid ();

  pid = fork ();
  if (pid < 0)
    {
      perror ("Could not fork");
    }
  if (pid > 0)
    {
      exit (EXIT_SUCCESS);
    }

  for (i = getdtablesize (); i >= 0; i--)
    {
      close (i);
    }

  i = open ("/dev/null", O_RDWR);
  dup (i);
  dup (i);

/*   umask (022); */
/*   chdir ("/tmp"); */
}
